src/ccl_power.c

```diff
@@ -873,120 +836,98 @@ static void ccl_cosmology_compute_power_eh(ccl_cosmology * cosmo, int * status)
+      } // end loop over k
+    } // end loop over a
   }
-
-  // Calculate sigma8 for the unnormalized P(k), using the standard
-  // ccl_sigma8() function
-  cosmo->data.p_lin = log_power_lin;
-  cosmo->computed_power = true; // Temporarily set this to true
-  double sigma8 = ccl_sigma8(cosmo, status);
-  cosmo->computed_power = false;
-
-  // Check that ccl_sigma8 didn't fail
-  if (*status) {
-    free(eh); free(x); free(y); free(a); free(y2d);
-    gsl_spline2d_free(log_power_lin);
-    return;
+  
+  // Check that ccl_growth_factor didn't fail
+  if (*status==0) {
+    cosmo->data.p_lin=ccl_p2d_t_new(na,z,nk,x,y2d,1,2,ccl_p2d_cclgrowth,1,NULL,0,ccl_p2d_3,status);
+    cosmo->computed_power=true;
+    sigma8 = ccl_sigma8(cosmo,status);
+    cosmo->computed_power=false;
   }
 
-  // Calculate normalization factor using computed value of sigma8, then
-  // recompute P(k, a) using this normalization
-  double log_normalization_factor = 2*(log(cosmo->params.sigma8) - log(sigma8));
-  for (int i=0; i < nk; i++) {
-    y[i] += log_normalization_factor;
+  if(*status==0) {
+    // Calculate normalization factor using computed value of sigma8, then
+    // recompute P(k, a) using this normalization
+    log_sigma8 = 2*(log(cosmo->params.sigma8) - log(sigma8));
+    for(int j=0;j<na*nk;j++)
+      y2d[j] += log_sigma8;
   }
-  for (int j = 0; j < na; j++) {
-    gfac = ccl_growth_factor(cosmo, a[j], status);
-    g2 = 2.*log(gfac);
-    for (int i=0; i<nk; i++) {
-      y2d[j*nk+i] = y[i]+g2; // Replace previous values
-    } // end k loop
-  } // end a loop
-
-  splinstatus = gsl_spline2d_init(log_power_lin, x, a, y2d, nk, na);
-  if (splinstatus) {
-    free(eh); free(x); free(y); free(a); free(y2d);
-    gsl_spline2d_free(log_power_lin);
-    *status = CCL_ERROR_SPLINE;
-    ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_eh(): Error creating log_power_lin spline\n");
-    return;
+  
+  if(*status==0) { 
+   // Free the previous P(k,a) spline, and allocate a new one to store the
+    // properly-normalized P(k,a)
+    ccl_p2d_t_free(cosmo->data.p_lin);
+    cosmo->data.p_lin=ccl_p2d_t_new(na,z,nk,x,y2d,1,2,ccl_p2d_cclgrowth,1,NULL,0,ccl_p2d_3,status);
   }
-  else
-    cosmo->data.p_lin = log_power_lin;
-
-  // Allocate a 2D spline for the nonlinear P(k) [which is just a copy of the
-  // linear one for E&H]
-  gsl_spline2d * log_power_nl = gsl_spline2d_alloc(PNL_SPLINE_TYPE, nk, na);
-  splinstatus = gsl_spline2d_init(log_power_nl, x, a, y2d, nk, na);
-
-  if (splinstatus) {
-    free(eh); free(x); free(y); free(a); free(y2d);
-    gsl_spline2d_free(log_power_lin);
-    gsl_spline2d_free(log_power_nl);
-    *status = CCL_ERROR_SPLINE;
-    ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_eh(): Error creating log_power_nl spline\n");
-    return;
+  
+  if(*status==0) {
+    if(cosmo->config.baryons_power_spectrum_method==ccl_bcm)
+      correct_bcm(cosmo,na,z,nk,x,y2d,status);
```

Comment:
are we applying BCM to the linear power spectrum here?

Reply:
We only apply it to whatever array is gonna end up in the non-linear power spectrum spline